exercises/linked-list/src/lib.rs

```diff
@@ -0,0 +1,74 @@
+pub struct LinkedList<T>(std::marker::PhantomData<T>);
+
+pub struct Cursor<'a, T: 'a>(std::marker::PhantomData<&'a mut T>);
+
+pub struct Iter<'a, T: 'a>(std::marker::PhantomData<&'a T>);
+
+impl<T> LinkedList<T> {
+    pub fn new() -> Self {
+        unimplemented!()
+    }
+
+    pub fn len(&self) -> usize {
+        unimplemented!()
+    }
+
+    /// Return a cursor positioned at the tail node
+    pub fn cursor_tail(&mut self) -> Cursor<'_, T> {
+        unimplemented!()
+    }
+
+    /// Return a cursor positioned at the head node
+    pub fn cursor_head(&mut self) -> Cursor<'_, T> {
+        unimplemented!()
+    }
+
+    pub fn iter(&self) -> Iter<'_, T> {
+        unimplemented!()
+    }
+}
+
+impl<T> Drop for LinkedList<T> {
+    fn drop(&mut self) {
+        unimplemented!()
+    }
+}
+
+// the cursor is expected to act as if it is at the position of a node
+impl<'a, T: 'a> Cursor<'a, T> {
+    /// Take a mutable reference to the current element
+    pub fn peek_mut(&mut self) -> Option<&mut T> {
```

Comment:
The existence of `peek_mut` implies `peek`, which is however not included in this API. I think that this is to save implementation time for the student; the mutable case is the more interesting case. I approve of this decision.

Reply:
The reason I left it out is that there's no use case for `peek`. The cursor is locked in place anyway and all other methods require `&mut`.

Reply:
Oh--I think I'd misread the way `Cursor` operated; I thought it mutably traversed the list.

If `Cursor` was changed to mutably traverse the list, and if `peek` were added, then it would be very simple to `impl Iterator for Cursor`. These changes may be worth making in order to reduce the total amount of implementation work required of the student; they wouldn't have to implement both `Cursor` and `Iter`. What do you think of this?

Reply:
I disagree. Let's work that example.

```rust
let first: &str = <Cursor as Iterator>::next(&mut cursor);
```

I'm pretty sure that `first` has to be `Option<&i32>` in this case. That's minor, though.

```rust
let first = <Cursor as Iterator>::next(&mut cursor);
```

`first` must equal `Some(&0)`, by the definition of the `Iterator` trait. It shouldn't borrow anything directly from `cursor`.

```rust
cursor.prev()
```

This should return `None`; the cursor is already on the first element. Let's pretend it said `cursor.next()` here, though: that advances the cursor and returns `Some(&mut 1)`, which reference is immediately destroyed.

```rust
cursor.take();
```

This returns `1` (immediately dropped) and removes its node. The `0` node should point forward to `2`, and the `2` node should point back at `0`. Nothing is dangling.

Reply:
> the cursor is already on the first element

This can't work. The iterator must progress on `next`, so it's now on the 2nd element.

Reply:
The `next` methods have different semantics. `Iterator::next`, called for the first time, returns the first element. `LinkedList::next`, called for the first time, advances the cursor and then returns the 2nd element.

Reply:
But both have to advance. What good is a Cursor iterator if it stays in place?

Reply:
I see. You're right. 

The upshot is that the borrow checker cannot allow `cursor.prev()` while `first` exists: there may not be a mutable reference and an immutable reference to the same data simultaneously. If the borrow checker forbids this code, then we still have a safe interface. 

An alternate solution would have the cursor's `prev` and `next` methods return `&T`, not `&mut T`. This should be fine, as we can still get a mutable reference by calling `peek_mut` if we need it. In that case the borrow checker should allow this code, and we still have a safe interface.

Reply:
There really is no way to do this without hacking weird state handling into the cursor. A structure that can mutate any node at any time and an iterator over references are incompatible interfaces.

Reply:
I think you're right. Our options are fundamentally either to restrict `Cursor` to immutability, or implement `Iter` as a separate struct.